CMakeFilters.cmake

```diff
@@ -85,30 +89,35 @@ endif ()
 if(NOT DEFINED ZLIBNG_PACKAGE_NAME)
   set(ZLIBNG_PACKAGE_NAME "zlib-ng")
 endif ()
-option (HDF5_ENABLE_ZLIB_SUPPORT "Enable Zlib Filters" OFF)
 if (HDF5_ENABLE_ZLIB_SUPPORT)
   if (NOT H5_ZLIB_HEADER)
     if (NOT ZLIB_USE_EXTERNAL)
+      option (HDF5_MODULE_MODE_ZLIB "Find the default system ZLIB in module mode" ON)
+      mark_as_advanced (HDF5_MODULE_MODE_ZLIB)
       if (HDF5_USE_ZLIB_NG)
+        set (HDF5_MODULE_MODE_ZLIB OFF CACHE BOOL "" FORCE)
         set (PACKAGE_NAME ${ZLIBNG_PACKAGE_NAME}${HDF_PACKAGE_EXT})
       else ()
         set (PACKAGE_NAME ${ZLIB_PACKAGE_NAME}${HDF_PACKAGE_EXT})
       endif ()
       set(ZLIB_FOUND FALSE)
-      if (HDF5_USE_ZLIB_STATIC)
-        set(ZLIB_SEARCH_TYPE static)
+      message (VERBOSE "Filter HDF5_ZLIB package name:${PACKAGE_NAME}")
+      if (HDF5_MODULE_MODE_ZLIB)
+        # Expect that the default shared library is expected with FindZLIB.cmake
+        find_package (ZLIB MODULE)
       else ()
-        set(ZLIB_SEARCH_TYPE shared)
-      endif ()
-      # Search pure Config mode first
-      find_package (ZLIB NAMES ${PACKAGE_NAME} OPTIONAL_COMPONENTS ${ZLIB_SEARCH_TYPE})
-      if (NOT ZLIB_FOUND)
-        if (CMAKE_VERSION VERSION_GREATER_EQUAL "3.24.0")
-          set(ZLIB_USE_STATIC_LIBS  ${HDF5_USE_ZLIB_STATIC})
-        endif()
-        find_package (ZLIB) # Legacy find
+        # Expect that a correctly built library with CMake config files is available
+        if (HDF5_USE_ZLIB_STATIC)
+          set(ZLIB_SEARCH_TYPE static)
+          if (CMAKE_VERSION VERSION_GREATER_EQUAL "3.24.0")
+            set(ZLIB_USE_STATIC_LIBS  ${HDF5_USE_ZLIB_STATIC})
+          endif()
+        else ()
+          set(ZLIB_SEARCH_TYPE shared)
+        endif ()
+        find_package (ZLIB NAMES ${PACKAGE_NAME} CONFIG OPTIONAL_COMPONENTS ${ZLIB_SEARCH_TYPE})
       endif ()
-      set(H5_ZLIB_FOUND ${ZLIB_FOUND})
+      set(H5_ZLIB_FOUND ZLIB_FOUND})
       if (H5_ZLIB_FOUND)
         if (HDF5_USE_ZLIB_NG)
```

Comment:
not related to the change directly, but isn't this wrong? zlib-ng can be built and found as libz.so and zlib.h too.